src/Functions/UserDefined/UserDefinedFunctionFactory.cpp

```diff
@@ -207,8 +212,36 @@ bool UserDefinedFunctionFactory::registerFunction(
 
         /// check whether the source can be compiled
         if (config->has("is_aggregation") && config->getValue<bool>("is_aggregation"))
+        {
             /// UDA
             V8::validateAggregationFunctionSource(function_name, {"initialize", "process", "finalize"}, config->get("source"));
+
+            /// add _tp_delta column as the last argument
+            assert(config->has("arguments"));
+            assert(config->isArray("arguments"));
+
+            bool has_delta_column = false;
+            /// Below is a workaround, because Poco::JSON::Object::getArray cannot work well
+            Poco::JSON::Array json_arguments = config->get("arguments").extract<Poco::JSON::Array>();
+            for (unsigned int i = 0; i < json_arguments.size(); i++)
+            {
+                const auto & arg = json_arguments.getObject(i);
+                if (arg->has("name") && arg->get("name") == "_tp_delta")
+                {
+                    has_delta_column = true;
+                    break;
+                }
+            }
+
+            if (!has_delta_column)
```

Comment:
I don't quite understand why we need `_tp_delta` for all cases.

Reply:
Because when writing source of UDA, the user never knows whether the UDA is used on a changelog stream or not,   he has to assume 'process' function always cope with _tp_delta column.

Reply:
Framework (proton) doesn't need to pass the delta column to UDA when in non-changelog scenario since this is more performant. The problem for now is in changelog scenarios, proton passes the delta column to UDA, can UDA differentiate these two cases from its inputs ?